crypto/schnorr-adaptor-signature/sig.go

```diff
@@ -1,271 +1,132 @@
+//  5. Let t be the byte-wise XOR of bytes(d) and tagged_hash("SchnorrAdaptor/aux", a)
+//     where a is auxiliary random data
+//  6. Let rand = tagged_hash("SchnorrAdaptor/nonce", t || T || P || m)
+func genRandForNonce(
+	skBytes [chainhash.HashSize]byte,
+	auxRand []byte,
+	encKeyBytes []byte,
+	pkBytes []byte,
+	msgHash []byte,
+) [chainhash.HashSize]byte {
+	// Step 5: t = bytes(d) XOR tagged_hash("SchnorrAdaptor/aux", aux)
+	auxHash := chainhash.TaggedHash(AdaptorSignatureTagAux, auxRand)
+	var t [chainhash.HashSize]byte
+	for i := 0; i < chainhash.HashSize; i++ {
+		t[i] = skBytes[i] ^ auxHash[i]
 	}
 
-	*sig = *adaptorSig
-
-	return nil
-}
-
-func (sig *AdaptorSignature) Equals(sig2 AdaptorSignature) bool {
-	return bytes.Equal(sig.MustMarshal(), sig2.MustMarshal())
-}
-
-// appendAndHash appends the given data and hashes the result
-// Expected input is:
-//   - msgHash: 32 bytes
-//   - signerPubKeyBytes: 33 bytes
-//   - encKeyBytes: 33 bytes
-//
-// The output is 32 bytes and is result of sha256(m || P || T)
-func appendAndHash(
-	msgHash []byte,
-	signerPubKeyBytes []byte,
-	encKeyBytes []byte,
-) []byte {
-	combinedData := make([]byte, 98)
-	copy(combinedData[0:32], msgHash)
-	copy(combinedData[32:65], signerPubKeyBytes)
-	copy(combinedData[65:98], encKeyBytes)
-	hash := sha256.Sum256(combinedData)
-	return hash[:]
+	// Step 6: rand = tagged_hash("SchnorrAdaptor/nonce", t || T || P || msg)
+	randForNonce := chainhash.TaggedHash(AdaptorSignatureTagNonce, t[:], encKeyBytes, pkBytes, msgHash)
+	return *randForNonce
 }
 
-// EncSign generates an adaptor signature by using the given secret key,
-// encryption key (noted by `T` in the paper) and message hash
+// EncSign creates an adaptor signature using the given private key, encryption key,
+// and message hash. It generates random auxiliary data internally.
 func EncSign(sk *btcec.PrivateKey, encKey *EncryptionKey, msgHash []byte) (*AdaptorSignature, error) {
-	// d' = int(d)
-	var skScalar btcec.ModNScalar
-	skScalar.Set(&sk.Key)
+	auxData := rand.Bytes(chainhash.HashSize)
+	return EncSignWithAuxData(sk, encKey, msgHash, auxData)
+}
 
+// EncSignWithAuxData creates an adaptor signature using the given private key,
+// encryption key, message hash, and auxiliary data.
+// allowing the caller to provide auxiliary data for deterministic nonce generation.
+func EncSignWithAuxData(sk *btcec.PrivateKey, encKey *EncryptionKey, msgHash []byte, auxData []byte) (*AdaptorSignature, error) {
 	// Fail if msgHash is not 32 bytes
 	if len(msgHash) != chainhash.HashSize {
 		return nil, fmt.Errorf("wrong size for message hash (got %v, want %v)", len(msgHash), chainhash.HashSize)
 	}
 
-	// Fail if d = 0 or d >= n
+	// Step 1: Let d' = int(d)
+	var skScalar btcec.ModNScalar
+	skScalar.Set(&sk.Key)
+
+	// Step 2: Return FAIL if d' == 0 or d' >= n
 	if skScalar.IsZero() {
 		return nil, fmt.Errorf("private key is zero")
 	}
 
-	// P = 'd*G
+	// Step 3: Let Pp = d' * G
 	pk := sk.PubKey()
```

Comment:
this function is not constant time as far as I can tell. you might want to use the blinded point multiplication routine instead.